### PR TITLE
test: Pr testing 

### LINKED ISSUE #6043: Allow setting of a global image registry override for the whole installer
This is a task derived from #5456 and it depends on #6042.

It is tricky to set all the right parameters in helm to override where the different keptn image should be pulled from since in the current installer, image and repository are both in the same variable.

This could be improved by introducing separate `image` and `repository` variables together also with image pull secrets that can be customized globally.
Those parameters are already provided globally from our helper functions in the common library helm chart.

Lots of work for this was already done in #5456 so it should be used as inspiration and baseline for this ticket.

### LINKED ISSUE #7578: Keptn Helm Chart flexibility for setting image repository and tag at one point
### Use Cases
- As a developer I want to test my keptn fork build artefacts in a local kubernetes cluster environment.
- As a developer I want to build and deploy my customized keptn fork at my kubernetes cluster environment.
- As a keptn contributor I want to fork keptn, build, deploy and test my enhancements locally in my kubernetes cluster before I open a Pull Request.
- As a mac book pro m1 (arm64) user, I want to build keptn locally and deploy it to my kubernetes cluster.
- As a developer I want to set the repository and image tag at one point for all keptn artefacts.

### Problem
All this use cases are depending on flexible and customizable helm charts. To be more specific, currently it is not possible to update the image repository and the image tag of the keptn services at one point. One has to go through different `values.yaml` and `Chart.yaml` files to hopefully find all the possible locations where to update the image repository and tag.

Example `installer/manifests/keptn/charts/control-plane/values.yaml`
```
shipyardController:
  image:
    repository: docker.io/keptn/shipyard-controller
    tag: ""
```
As we can see here, the repository and the tag are hardcoded properties. As there are a lot of services in keptn where it is necessary to update the repository and image tag, it is a huge source of error.

## Solution & Proposal
We can use the helm templating engine to set the desired image repository and image tag in `values.yaml` and `values-local.yaml` at one point. If we introduce global values, we can achieve a much higher flexibility.

Example
`/installer/manifests/keptn/charts/control-plane/values.yaml`
```
global:
  keptn:
    registry: docker.io/keptn 
    tag: "" 

...
apiService:
  ...
  image:
    registry: "" 
    repository: "api" 
    tag: "" 

distributor:
  ...
  image:
    registry: ""
    repository: "distributor"
    tag: "" 
```
`installer/manifests/keptn/charts/control-plane/templates/api-service.yaml`
```
...
app.kubernetes.io/version: {{ include "service.tag" ( dict "imageRoot" .Values.apiService.image "global" .Values.global.keptn "defaultTag" .Chart.AppVersion) }}
...
image: {{ include "service.image" ( dict "imageRoot" .Values.apiService.image "global" .Values.global.keptn "defaultTag" .Chart.AppVersion) }}
...
```
Optional: One can use a values-local.yaml for customizing the repository and the appVersion
```
global:
  keptn:
    registry: "test/it/like-a-dude"
    appVersion: mylocalVersion
```
Introducing a new helper template `_images.tpl` in `installer/manifests/common` in bitnami style, helps us to decide about the registry, repository or image tag.

In addition to the changes above, the `installer/airgapped/install_keptn.sh` script can be fully refactored and simplified. For example: one would not have to add the registry (repository) name as an input parameter but can only set a flag like `values-local.yaml=true`. 

The benefit of this feature would be, the charts are open for new future custom configurations. As the only thing a developer would have to do, is to add a values-local.yaml. 

This proposal has no incompatibilities within the current keptn helm charts but rather introduces a flexible extension point to use custom keptn builds.

I am open for feedback as well as for other solutions!

### PR DESCRIPTION
**Pull request by RealAnna**
## This PR

- fixes a few bugs in #8152

### Related Issues

Closes  #8152
Fixes #7578
Fixes #6043

### How to test

https://github.com/keptn/keptn/actions/runs/2620873127

### CODE REVIEW

File: .github/workflows/integration_tests.yml [hunk 1]
```diff
@@ -352,68 +352,26 @@ jobs:
...
-              --set continuousDelivery.enabled=${RUN_CONTINUOUS_DELIVERY_TEST} \
-              --wait \
-              --timeout=10m
+            if [[ "$HELM_CHART_NAME" == *"dev"* ]]; then
+              helm install -n ${KEPTN_NAMESPACE} keptn ${HELM_CHART_NAME} \
+                -f ./test/assets/helm-charts/keptn/values-int-tests.yaml \
+                --create-namespace \
+                --set apiGatewayNginx.type=${KEPTN_SERVICE_TYPE} \
+                --set shipyardController.gracePeriod=10 \
+                --set shipyardController.preStopHookTime=5 \
```
Comment (mowies) [inline .github/workflows/integration_tests.yml:361]: You could put those hardcoded values directly into the `values-int-tests` file

File: .github/workflows/integration_tests.yml [hunk 2]
```diff
@@ -352,68 +352,26 @@ jobs:
...
+                --set continuousDelivery.enabled=${RUN_CONTINUOUS_DELIVERY_TEST} --wait --timeout 12m
+            else
+              helm install -n ${KEPTN_NAMESPACE} keptn ${HELM_CHART_NAME} \
+                --create-namespace \
```
Comment (mowies) [inline .github/workflows/integration_tests.yml:365]: Why don't we need the `values-int-tests` file here?

File: .github/workflows/integration_tests.yml [hunk 3]
```diff
@@ -462,35 +420,126 @@ jobs:
           if [[ "$REMOTE_EXECUTION_PLANE" == 'true' ]]; then
             # Remote execution plane
             KEPTN_API_TOKEN=$(kubectl get secret keptn-api-token -n "$KEPTN_NAMESPACE" -o jsonpath='{.data.keptn-api-token}' | base64 --decode)
-          
+            echo "remote"
```
Comment (mowies) [inline .github/workflows/integration_tests.yml:423]: I guess this can be removed now :)

File: .github/workflows/integration_tests.yml [hunk 4]
```diff
@@ -462,35 +420,126 @@ jobs:
...
-            helm install jmeter-service "${JMETER_SERVICE_HELM_CHART_NAME}" \
-            -n ${{ env.KEPTN_NAMESPACE }}-jmeter-service \
-              --set remoteControlPlane.enabled=true \
-              --set remoteControlPlane.api.protocol=http \
-              --set remoteControlPlane.api.hostname="${KEPTN_API_HOSTNAME}" \
-              --set remoteControlPlane.api.token="${KEPTN_API_TOKEN}" \
-              --create-namespace
+            if [[ "$HELM_SERVICE_HELM_CHART_NAME" == *"dev"* ]]; then
+              echo "installing helm service from dev repo"
+              echo 'helm install helm-service ${HELM_SERVICE_HELM_CHART_NAME} --namespace ${{ env.KEPTN_NAMESPACE }}-helm-service --set remoteControlPlane.enabled=true --set remoteControlPlane.api.protocol=http --set remoteControlPlane.api.hostname="${KEPTN_API_HOSTNAME}" --set remoteControlPlane.api.token="${KEPTN_API_TOKEN}" --create-namespace --values "./test/assets/helm-charts/helm-service/values-int-tests.yaml"'
```
Comment (mowies) [inline .github/workflows/integration_tests.yml:428]: I guess this can be removed now :)

File: .github/workflows/integration_tests.yml [hunk 5]
```diff
@@ -462,35 +420,126 @@ jobs:
...
+              helm install helm-service ${HELM_SERVICE_HELM_CHART_NAME} --namespace ${{ env.KEPTN_NAMESPACE }}-helm-service \
+                --set remoteControlPlane.enabled=true \
+                --set remoteControlPlane.api.protocol=http \
+                --set remoteControlPlane.api.hostname="${KEPTN_API_HOSTNAME}" \
+                --set remoteControlPlane.api.token="${KEPTN_API_TOKEN}" \
+                --create-namespace \
+                --values "./test/assets/helm-charts/helm-service/values-int-tests.yaml"
+
+            else
+              echo "installing helm service from keptn repo"
```
Comment (mowies) [inline .github/workflows/integration_tests.yml:438]: I guess this can be removed now :)

File: .github/workflows/integration_tests.yml [hunk 6]
```diff
@@ -462,35 +420,126 @@ jobs:
...
+                --set remoteControlPlane.enabled=true \
+                --set remoteControlPlane.api.protocol=http \
+                --set remoteControlPlane.api.hostname="${KEPTN_API_HOSTNAME}" \
+                --set remoteControlPlane.api.token="${KEPTN_API_TOKEN}" \
+                --create-namespace \
+                --values "./test/assets/helm-charts/helm-service/values-int-tests.yaml"
+
+            else
+              echo "installing helm service from keptn repo" 
+              helm install helm-service ${HELM_SERVICE_HELM_CHART_NAME} -n ${{ env.KEPTN_NAMESPACE }}-helm-service \
```
Comment (mowies) [inline .github/workflows/integration_tests.yml:439]: No values file here?

File: .github/workflows/integration_tests.yml [hunk 7]
```diff
@@ -462,35 +420,126 @@ jobs:
...
+                --set remoteControlPlane.enabled=true \
+                --set remoteControlPlane.api.protocol=http \
+                --set remoteControlPlane.api.hostname="${KEPTN_API_HOSTNAME}" \
+                --set remoteControlPlane.api.token="${KEPTN_API_TOKEN}" \
+                --create-namespace
+            fi
+          
+            if [[ "$JMETER_SERVICE_HELM_CHART_NAME" == *"dev"* ]]; then
+              echo "installing jmeter from dev repo"
```
Comment (mowies) [inline .github/workflows/integration_tests.yml:448]: I guess this can be removed now :)

File: .github/workflows/integration_tests.yml [hunk 8]
```diff
@@ -462,35 +420,126 @@ jobs:
...
+                --set remoteControlPlane.enabled=true \
+                --set remoteControlPlane.api.protocol=http \
+                --set remoteControlPlane.api.hostname="${KEPTN_API_HOSTNAME}" \
+                --set remoteControlPlane.api.token="${KEPTN_API_TOKEN}" \
+                --create-namespace \
+                --values "./test/assets/helm-charts/jmeter-service/values-int-tests.yaml"
+                
+            else
+              echo "installing jmeter from keptn repo"
+              helm install jmeter-service ${JMETER_SERVICE_HELM_CHART_NAME} -n ${{ env.KEPTN_NAMESPACE }}-jmeter-service \
```
Comment (mowies) [inline .github/workflows/integration_tests.yml:459]: no values file here?

File: .github/workflows/integration_tests.yml [hunk 9]
```diff
@@ -462,35 +420,126 @@ jobs:
...
-              --set remoteControlPlane.api.token="${KEPTN_API_TOKEN}" \
-              --create-namespace
+            if [[ "$HELM_SERVICE_HELM_CHART_NAME" == *"dev"* ]]; then
+              echo "installing helm service from dev repo"
+              echo 'helm install helm-service ${HELM_SERVICE_HELM_CHART_NAME} --namespace ${{ env.KEPTN_NAMESPACE }}-helm-service --set remoteControlPlane.enabled=true --set remoteControlPlane.api.protocol=http --set remoteControlPlane.api.hostname="${KEPTN_API_HOSTNAME}" --set remoteControlPlane.api.token="${KEPTN_API_TOKEN}" --create-namespace --values "./test/assets/helm-charts/helm-service/values-int-tests.yaml"'
+              helm install helm-service ${HELM_SERVICE_HELM_CHART_NAME} --namespace ${{ env.KEPTN_NAMESPACE }}-helm-service \
+                --set remoteControlPlane.enabled=true \
+                --set remoteControlPlane.api.protocol=http \
+                --set remoteControlPlane.api.hostname="${KEPTN_API_HOSTNAME}" \
+                --set remoteControlPlane.api.token="${KEPTN_API_TOKEN}" \
```
Comment (mowies) [inline .github/workflows/integration_tests.yml:433]: maybe a separate remote exec plane values file would make sense to cut down on pipeline LoC?

File: installer/manifests/common/templates/_images.tpl [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+{{- if $registryName }}
+{{- printf "%s/%s:%s" $registryName $repositoryName $tag -}}
+{{- else -}}
+{{- printf "%s:%s" $repositoryName $tag -}}
+{{- end -}}
+{{- end -}}
+
+{{/*Return the proper serivce image tag*/}}
+{{/*{{- include "common.images.tag" ( dict "imageRoot" .Values.helmservice.image "global" .Values.global.keptn "defaultTag" .Chart.appVersion) -}}*/}}
+{{- define "common.images.tag" -}}
```
Comment (mowies) [inline installer/manifests/common/templates/_images.tpl:24]:
```suggestion
{{/*{{- include "keptn.common.images.tag" ( dict "imageRoot" .Values.helmservice.image "global" .Values.global.keptn "defaultTag" .Chart.appVersion) -}}*/}}
{{- define "keptn.common.images.tag" -}}
```

File: installer/manifests/common/templates/_images.tpl [hunk 11]
```diff
@@ -0,0 +1,39 @@
+{{/*Return the proper serivce image name*/}}
+{{/*{{- include "common.images.image" ( dict "imageRoot" .Values.helmservice.image "global" .Values.global.keptn "defaultTag" .Chart.appVersion) -}}*/}}
+{{- define "common.images.image" -}}
+{{- $registryName := "" -}}
+{{- $repositoryName := .imageRoot.repository -}}
+{{- $tag := include "common.images.tag" (dict "imageRoot" .imageRoot "global" .global "defaultTag" .defaultTag) -}}
```
Comment (mowies) [inline installer/manifests/common/templates/_images.tpl:6]:
```suggestion
{{- $tag := include "keptn.common.images.tag" (dict "imageRoot" .imageRoot "global" .global "defaultTag" .defaultTag) -}}
```

File: installer/airgapped/install_keptn.sh [hunk 12]
```diff
@@ -19,29 +19,25 @@ echo "-----------------------------------------------------------------------"
...
-configurationService.image.repository=${TARGET_INTERNAL_DOCKER_REGISTRY}${DOCKER_ORG}/configuration-service,\
-resourceService.image.repository=${TARGET_INTERNAL_DOCKER_REGISTRY}${DOCKER_ORG}/resource-service,\
-mongodbDatastore.image.repository=${TARGET_INTERNAL_DOCKER_REGISTRY}${DOCKER_ORG}/mongodb-datastore,\
-statisticsService.image.repository=${TARGET_INTERNAL_DOCKER_REGISTRY}${DOCKER_ORG}/statistics-service,\
-lighthouseService.image.repository=${TARGET_INTERNAL_DOCKER_REGISTRY}${DOCKER_ORG}/lighthouse-service,\
-secretService.image.repository=${TARGET_INTERNAL_DOCKER_REGISTRY}${DOCKER_ORG}/secret-service,\
-approvalService.image.repository=${TARGET_INTERNAL_DOCKER_REGISTRY}${DOCKER_ORG}/approval-service,\
-webhookService.image.repository=${TARGET_INTERNAL_DOCKER_REGISTRY}${DOCKER_ORG}/webhook-service"
+apiGatewayNginx.image.registry="",\
+apiGatewayNginx.image.repository=${TARGET_INTERNAL_DOCKER_REGISTRY}nginxinc/nginx-unprivileged"
```
Comment (mowies) [inline installer/airgapped/install_keptn.sh:40]: Why don't you use the `registry` value to set it to `${TARGET_INTERNAL_DOCKER_REGISTRY}` and then use the `repository` value to set it to `nginxinc/nginx-unprivileged`?

File: .github/workflows/integration_tests.yml (hunk 6 above)
Comment (RealAnna) [reply .github/workflows/integration_tests.yml:439]: value file only override repo to dev so here no

File: .github/workflows/integration_tests.yml (hunk 8 above)
Comment (RealAnna) [reply .github/workflows/integration_tests.yml:459]: same as helm service

File: test/assets/helm-charts/keptn/values-int-tests.yaml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+global:
+  keptn:
+    registry: "docker.io/keptn"
+
+shipyardController:
+  gracePeriod: 10
+  preStopHookTime: 5
```
Comment (mowies) [inline test/assets/helm-charts/keptn/values-int-tests.yaml:7]: pls add the missing newline :)

File: test/assets/helm-charts/keptn/values-int-tests-dev.yaml [hunk 14]
```diff
@@ -0,0 +1,7 @@
+global:
+  keptn:
+    registry: "docker.io/keptndev"
+
+shipyardController:
+  gracePeriod: 10
+  preStopHookTime: 5
```
Comment (mowies) [inline test/assets/helm-charts/keptn/values-int-tests-dev.yaml:7]: pls add the missing newline :)

**Pull request closed without merging**